rcl/src/rcl/node.c

```diff
@@ -121,13 +122,27 @@ rcl_node_init(
   rcl_context_t * context,
   const rcl_node_options_t * options)
 {
+  const char * env_error_str = NULL;
+  const char * env_val = NULL;
+
   const rmw_guard_condition_t * rmw_graph_guard_condition = NULL;
   rcl_guard_condition_options_t graph_guard_condition_options =
     rcl_guard_condition_get_default_options();
   rcl_ret_t ret;
   rcl_ret_t fail_ret = RCL_RET_ERROR;
   char * remapped_node_name = NULL;
 
+  env_error_str = rcutils_get_env(RCL_NAMESPACE_ENV_VAR, &env_val);
+  if (NULL != env_error_str) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting env var: '" RCUTILS_STRINGIFY(RCL_NAMESPACE_ENV_VAR) "': %s\n",
+      env_error_str);
+    return RCL_RET_ERROR;
+  }
+  if (NULL != env_val && 0 != strlen(env_val)) {
+    namespace_ = env_val;
+  }
```

Comment:
I think the second option is the most viable one. In ROS2 there may be really complex launch files, and I may need to add additional namespaces while I am launching them. Doing every namespace declaration again in every launch might be too cumbersome. A good solution may be giving an input to the launch file, and deciding the namespace accordingly, but still just giving an environment variable is simpler.

Reply:
A quick empirical test shows that ROS 1 prepends the value of `ROS_NAMESPACE` _only_ if the node's namespace is not explicitly absolute (at least in the context of a launch file):

With this launch file:
```xml
<launch>
  <node name="talker" ns="bar" pkg="rospy_tutorials" type="talker" />
</launch>
```
and `ROS_NAMESPACE=/foo`, we get the node name `/foo/bar/talker`:

But, with this launch file (note the absolute namespace), we get the node name `/bar/talker`:

```xml
<launch>
  <node name="talker" ns="/bar" pkg="rospy_tutorials" type="talker" />
</launch>
```

I don't know how easy it will be to implement the same behavior in ROS 2. Note, launch files in ROS 2 deal with namespaces by passing remaps to the node process. So, if we want to prepend the value of `ROS_NAMESPACE`, I think we have to do that _after_ processing any remaps.

Correct me if I'm wrong, but I think `namespace_` is the name given in code (ie. by the node author). So as currently implemented, `ROS_NAMESPACE` takes affect _before_ any name remapping.

Reply:
Replicating the ROS1 behaviour seems the best choice to me. If no one disagrees, I may look at how hard it will be.

Reply:
> Replicating the ROS1 behaviour seems the best choice to me. If no one disagrees, I may look at how hard it will be.

+1 as well.
Could you update the code to work like that?

Reply:
BTW, should the environment variable be `ROS_NAMESPACE` instead?
e.g. for the domain id we use `ROS_DOMAIN_ID`.
@clalancette what do you think?

Reply:
> BTW, should the environment variable be ROS_NAMESPACE instead?

Yes, it should be `ROS_NAMESPACE`.  It looks like the latest code changes to use that, so I think at least that part of it is OK.